### ISSUE TITLE
Stuck on findOne

### ISSUE THREAD

**Issue by Christilut** [szokodiakos/typegoose#64]
I probably made a mistake implementing Typegoose somewhere (btw, awesome project!) but on the first `User.findOne()` my request is freezing.

This is my (simplified) model:
```ts
export class UserClass extends Typegoose {
  @prop()
  _id: string

  @prop({
    required: true
  })
  name: string

  @prop({
    required: true
  })
  email: string

  @prop({
    required: true
  })
  password: string
}
```
Exposed like this:
```ts
export const User = new UserClass().getModelForClass(UserClass)
```

And called like this:

```ts
import { UserClass, User } from 'server/models/user/model'

  const user = await User.findOne({
    email
  })
```

But it gets stuck on the `User.findOne` line and never returns from it.

Last thing the console prints before that is `DeprecationWarning: Mongoose: mpromise (mongoose's default promise library) is deprecated` even though I'm using `(mongoose as any).Promise = Promise`.
Before Typegoose I didn't get the deprecation warning and the Promise line did work.

Typegoose 3.6.0
Node v8.5.0

Any ideas?

**Comment by Christilut** [comment]
Probably a mongoose issue but I really can't figure this one out.

Tried bluebird, older mongoose versions, lots of different settings but nothing seems to work.

**Comment by rasdaniil** [comment]
Same story! If I switch to native mongoose schema way - everything works :-(

**Comment by rasdaniil** [comment]
Tried to supply 'existingCollection' to models - doesn't help

**Comment by Christilut** [comment]
I'm using typegoose to create a boilerplate so anyone that wants to take look: [here it is](https://github.com/Christilut/node-api-boilerplate/tree/typegoose-hangs)

Use the `typegoose-hangs` branch.

Copy `.env.example` to `.env`, run `npm install`, then `npm run dev` and it should work (I hope :stuck_out_tongue:)

This is the last I see:
```
saving new user for testing
pre save !!!
(node:1685) DeprecationWarning: Mongoose: mpromise (mongoose's default promise library) is deprecated, plug in your own promise library instead: http://mongoosejs.com/docs/promises.html
```

Check `/node-api-boilerplate/server/models/user/model.ts` for relevant model code and `config/mongoose.ts` for mongoose init.

**Comment by rasdaniil** [comment]
The problem is that typegoose has mongoose (and typescript) defined as dependencies in package.json. And versions are set strict. That's why typegoose actually brings his own mongoose and typescript inside node_modules/typegoose/node_modules , and uses not that mongoose which we require and install separately

This deps should be peerDependencies or maybe the version should not be set strict

:-(

**Comment by rasdaniil** [comment]
This is a workaround

```
yarn remove typegoose mongoose
yarn add mongoose@4.11.1 typegoose

```

**Comment by Christilut** [comment]
Nice work! Removing `node_modules/typegoose/node_modules/mongoose` fixed it for me (as a quick test).

And your above workaround works too.

**Comment by Christilut** [comment]
Just noticed that my pre-save stopped working after this..

```ts
@pre<User>('save', async function (next) {
  console.log('pre save !!!') 

  next()
})

export class User extends Typegoose {
}

export const UserModel = new User().getModelForClass(User)
```

This is pretty much the example code (and it did work earlier).

**Comment by Christilut** [comment]
My pre-save suddenly started working again, this is what happenend:

In my model I had the `created` property, which is to be set in the pre-save
```ts
  @prop({
    required: true
  })
  created: Date
```

Pre-save:
```ts
  if (this.isNew) {
    this.created = new Date()
  }
```

But that triggered a validation error that `created` was required. So apparently validation is triggered before the pre-save and properties that get set in the pre-save shouldn't have `required` on them.

**Comment by vinczedani** [comment]
Hi @Christilut 
We are aware of the issue that typegoose has its own mongoose version. It is on our todo list to list mongoose as a peer dependency instead of a normal one. It should resolve some issues.

The pre save is after the validation, but mongoose offers pre validation hook too. In your case, you should give it a try ;)

**Comment by valneras** [comment]
Hello,
I was also blocked during a few hours because of the blocking error :(

> DeprecationWarning: Mongoose: mpromise (mongoose's default promise library) is deprecated

I also removed manually the folder "node_modules/typegoose/node_modules/mongoose" as a workaround and it has worked.

However is there another proper way to avoid this ?

Thanks!

**Comment by neckaros** [comment]
very hard to find issue indeed! my await was just never returning and nerver throwing!

**Comment by vinczedani** [comment]
Hi @valneras 

You should pass your own mongoose instance in the getModelForClass method if you have any. The second parameter (options) has a property called `existingMongoose`. This should make the typegooses owned fallback mongoose unused, so the deprecation warning should disappear.

You can also pass in an existing mongoose connection if you have any. It can also passed down to typegoose in the getModelForClass methods second parameter. Under the `existingConnection` key in the options object.

The precedency of all the mongooses inside typegoose is the following:
1. existingConnection
2. existingMongoose
3. typegooses own mongoose

Unfortunately, we don't really have spare time for typegoose at the moment. In the future we plan to deprecate our own mongoose as a dependency and add it as a peer dependency instead. But since it is a breaking change for those who rely on the built in mongoose instance, it is more complicated to do, than just changing two lines in the package.json.
I hope this will fix your problem.

Cheers,
Daniel

**Comment by hasezoey** [comment]
this can be closed, because mongoose 5.0+ uses the default `Promise`, and typegoose uses an "up-to-date" mongoose version

@Ben305
